<a href="https://colab.research.google.com/github/aserdargun/aserdargun/blob/main/MyPythonDeepDive.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# INTRODUCTION

I am starting with the zen of python. If you write "import this" and understand what is this, you will see significant effects on your syntax.

In [ ]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


If you want to have more details on it, you should go https://peps.python.org/pep-0008/ PEP8. PEP is an abbrevation of Python Enchancement Proposals. 

You can see the PEP index in this https://peps.python.org/pep-0000/

The resource of this study is from Fred Baptiste's DeepDive Series. Special thanks to Fred Baptiste. https://github.com/aserdargun/python-deepdive

* [Part 1](https://www.udemy.com/course/python-3-deep-dive-part-1/?referralCode=E46B931C71EE01845062/): Mainly functional programming
* [Part 2](https://www.udemy.com/course/python-3-deep-dive-part-2/?referralCode=3E7AFEF5174F04E5C8D4/): Mainly iterables, iterators and generators
* [Part 3](https://www.udemy.com/course/python-3-deep-dive-part-3/?referralCode=C5B0D9AB965B9BF4C49F/): Mainly hash maps
* [Part 4](https://www.udemy.com/course/python-3-deep-dive-part-4/?referralCode=3BB758BE4C04FB983E6F/): OOP


I only want to emphasize some caveats, also good understanding for me and sharing with others who want to learn Python in details.

# **PART 1: FUNCTIONAL PROGRAMMING**


## Section 02 - A Quick Refresher

### 01 - Multi-Line Statements and Strings

**Implicit Examples**

Multi-Line Statement

In [ ]:
a = [1,
     2,
     3]

In [ ]:
a

[1, 2, 3]

Multi-Line Statement with comments to the end of each line.

In [ ]:
a = [1, #first element
     2, #second element
     3, #third element
     ]

In [ ]:
a

[1, 2, 3]



---
**BE CAREFUL!**


*If you use comments, you must close off the collection on a new line.*

In [ ]:
a = [1, #first element
     2 # second element]

SyntaxError: ignored



---



This works the same way for tuples, sets, and dictionaries.

Tuple

In [ ]:
a = (1, #first element
     2, #second element
     3, #third element
     )

In [ ]:
a

(1, 2, 3)

Set

In [ ]:
a = {1, #first element
     2, #second element
     }

In [ ]:
a

{1, 2}

Dictionary

you can use each of key and value pairs comments in dictionary.

In [ ]:
a = {'key1': 'value1', #comment,
     'key2': #comment
     'value2' #comment
     }

In [ ]:
a

{'key1': 'value1', 'key2': 'value2'}

Function Parameter

A parameter is the variable listed inside the parentheses in the function definition.

In [ ]:
def my_func(a, #some comment
            b, c):
    print(a, b ,c)

Function Argument

An argument is the value that are sent to the function when it is called.

In [ ]:
my_func(10, #comment
        20, #comment
        30)

10 20 30


**Explicit Examples**

You can use the `\` character to explicitly create multi-line statements.

In [ ]:
a = 10
b = 20
c = 30
if a > 5 \
  and b > 10 \
  and c > 20:
  print('yes!!')

yes!!


The indentation in continued-lines does not matter:

In [ ]:
a = 10
b = 20
c = 30
if a > 5 \
    and b > 10 \
        and c > 20:
    print('yes!!')

yes!!


**Multi-Line Strings**

You can create multi-line strings by using triple delimiters (single or double quotes)

In [ ]:
a = '''this is 
a multi-line string'''

In [ ]:
print(a)

this is 
a multi-line string


If you call with a directly in notebook, you will see with escape characters such as `\n`.

In [ ]:
a

'this is \na multi-line string'

Any character you type is preserved. You can also mix in escaped characters line any nırmal string. (i.e.: `\n`)

In [ ]:
a = """some items:\n
    1. item 1
    2. item 2"""

In [ ]:
print(a)

some items:

    1. item 1
    2. item 2




---
**BE CAREFUL!**


*if you indent your multi-line strings - the extra spaces are preserved!*

In [ ]:
def my_func():
    a = '''a multi-line string
    that is actually indented in the second line'''
    return a

In [ ]:
print(my_func())

a multi-line string
    that is actually indented in the second line


In [ ]:
def my_func():
    a = '''a multi-line string
that is not indented in the second line'''
    return a

In [ ]:
print(my_func())

a multi-line string
that is not indented in the second line




---

**BE CAREFUL!**

*Note that these multi-line strings are not comments - they are real strings and, unlike comments, are part of your compiled code. They are however used to create comments, such as docstrings, that we will cover later in this course.*



---

**BE CAREFUL!**

*In general, use `#` to comment your code, and use multi-line strings only when actually needed (like for docstrings).
Also, there are no multi-line comments in Python. You simply have to use a `#` on every line.*

In [ ]:
# this is
#   a multi-line
#   comment

*The following works, but the above formatting is preferrable.*

In [ ]:
# this is
    # a multi-line
    # comment



---



### 02 - Conditionals

This is achieved using `if`, `elif` and `else` or the ternary operator (aka conditional expression)

In [ ]:
a = 2
if a < 3:
    print('a < 3')
else:
    print('a >= 3')

a < 3


`if` statements can be nested.

In [ ]:
a = 15

if a < 5:
    print('a < 5')
else:
    if a < 10:
        print('5 <= a < 10')
    else:
        print('a >= 10')

a >= 10


But the `elif` statement provides for better readability.

In [ ]:
a = 15

if a < 5:
    print('a < 5')
elif a < 10:
    print('5 <= a < 10')
else:
    print('a >= 10')

a >= 10


Python also provides a conditional expression (ternary operator):

`X if (condition) else Y`

returns (and evaluates) X if (conditions) is True, otherwise returns (and evaluates) Y

In [ ]:
a = 5
res = 'a < 10' if a < 10 else 'a >= 10'
print(res)

a < 10


In [ ]:
a = 15
res = 'a < 10' if a < 10 else 'a >= 10'
print(res)

a >= 10




---
**BE CAREFUL!**


*Note that X and Y can be any expression, not just literal values:*

In [ ]:
def say_hello():
    print('Hello!')

def say_goodbye():
    print('Goodbye!')

In [ ]:
a = 5
say_hello() if a < 10 else say_goodbye()

Hello!


In [ ]:
a = 15
say_hello() if a < 10 else say_goodbye()

Goodbye!




---



### 03 - Functions

Python has many built-in functions and methods we can use
Some are available by default:

In [ ]:
s = [1, 2, 3]
len(s)

3

While some need to be imported:

In [ ]:
from math import sqrt

In [ ]:
sqrt(4)

2.0

Entire modules can be imported:

In [ ]:
import math

In [ ]:
math.exp(1)

2.718281828459045

We can define our own functions:

In [ ]:
def func_1():
    print('running func1')

In [ ]:
func_1()

running func1




---
**BE CAREFUL!**


*Note that to "call" or "invoke" a function we need to use the `()`.
Simply using the function name without the `()` refers to the function, but does not call it:* 

In [ ]:
func_1

<function __main__.func_1()>



---



*We can also define functions that take parameters:*

In [ ]:
def func_2(a, b):
    return a * b



---
**BE CAREFUL!**


*Note that `a` and `b` can be any type (this is an example of polymorphism - which we will look into more detail later in this course).
But the function will fail to run if `a` and `b` are types that are not "compatible" with the ` * ` operator:*

In [ ]:
func_2(3, 2)

6

In [ ]:
func_2('a', 3)

'aaa'

In [ ]:
func_2('a', 'b')

TypeError: ignored



---



It is possible to use type annotations:

In [ ]:
def func_3(a: int, b: int):
    return a * b

In [ ]:
func_3(2 ,3)

6

In [ ]:
func_3('a', 2)

'aa'



---
**BE CAREFUL!**


*But as you can see, these do not enforce a data type! They are simply metadata that can be used by external libraries, and many IDE's.*



---



Functions are objects, just like integers are objects, and they can be assigned to variables just as an integer can:

In [ ]:
my_func = func_3

In [ ]:
my_func('a', 2)

'aa'



---
**BE CAREFUL!**


*Functions must always return something. If you do not specify a return value, Python will automatically return the `None` object:*

In [ ]:
def func_4():
    # does something but does not return a value
    a = 2

In [ ]:
res = func_4()

In [ ]:
print(res)

None




---



The `def` keyword is an executable piece of code that creates the function (an instance of the function class) and essentially assigns it to a variable name ( the function name).



---
**BE CAREFUL!**


*Note that the function is defined when `def` is reached, but the code inside it is not evaluated until the function is called.
This is why we can define functions that call other functions defined later - as long as we don't call them before all the necessary functions are defined.
For example, the following will work:*

In [ ]:
def fn_1():
    fn_2()

def fn_2():
    print('Hello')

fn_1()

Hello


*But this will not work:*

In [ ]:
def fn_3():
    fn_4

fn_3()

def fn_4():
    print('Hello')

NameError: ignored



---



We also have the `lambda` keyword, that also creates a new function, but does not assign it to any specific name - instead it just returns the function object - which we can, if we wish, assign yo a variable ourselves:

In [ ]:
func_5 = lambda x: x ** 2

In [ ]:
func_5

<function __main__.<lambda>(x)>

In [ ]:
func_5(2)

4

### 04 - The While Loop

The while loop is a way to repeat a block of code as long as a specified condition is met.

```
while <exp is true>:
    code block
```

In [ ]:
i = 0
while i < 5:
    print(i)
    i += 1

0
1
2
3
4




---
**BE CAREFUL!**


*Note that there is no guarantee that a while loop will execute at all, not even once, because the condition is tested before the loop runs.*

In [ ]:
i = 5
while i < 5:
    print(i)
    i += 1

*Some languages have a concept of a while loop that is guaranteed to executed at least one:*
```
do
    code block
while <exp is true>
```
*There is no such thing in Python, but it's easy to write code that works that way.*




---



We create an infinite loop and test the condition inside the loop and break out of the loop when the condition becomes false:

In [ ]:
i = 5

while True:
    print(i)
    if i >= 5:
        break

5


As you can see the loop executed once ( and will always execute at least once, no matter the starting value of `i`.)
This is a standard pattern and can be useful in a varity of scenarios.
A simple example might be getting repetitive user input until the user performs and action or provides some specific value.
we might try it this way:

In [ ]:
min_length = 2
name = input('Please enter your name:')
while not(len(name) >= min_length and name.isprintable() and name.isalpha()):
    name = input('Please enter your name:')
print('Hello, {0}'.format(name))

Please enter your name:a123
Please enter your name:a
Please enter your name:serdar
Hello, serdar


This works just fine, but notice that we had to write the code to elicit user input twice in our code. This is not good practice, and we can easily clean this up as follows:

In [ ]:
min_length = 2

while True:
    name = input('Please enter your name:')
    if len(name) >= min_length and name.isprintable() and name.isalpha():
        break

print('Hello, {0}'.format(name))

Please enter your name:a
Please enter your name:123
Please enter your name:serdar
Hello, serdar


We saw how the `break` statement exists the while loop and execution resumes on the line immediately after the while code block.

Sometimes, we just want to cut the current iteration short, but continue looping, without exiting the loop itself.

This is done using `continue` statement.

In [ ]:
a = 0
while a < 10:
    a += 1
    if a % 2:
        continue
    print(a)

2
4
6
8
10




---
**BE CAREFUL!**


*The while loop also can be used with an `else` clause!!*

*The `else` is executed if the `while` loop terminated without hitting a `break` statement (we say the loop terminated normally)*

*Suppose we want to test if some value is present in some list, and if not we want to append it to the list (again there are better ways of doing this):*

*First, here's how we might do it wihout the benefit of the `else` clause:*

In [ ]:
l = [1, 2, 3]
val = 10

found = False
idx = 0
while idx < len(l):
    if l[idx] == val:
        found = True
        break
    idx += 1

if not found:
    l.append(val)
print(l)

[1, 2, 3, 10]


Using the `else` clause is easier:

In [ ]:
l = [1, 2, 3]
val = 10

idx = 0
while idx < len(l):
    if l[idx] == val:
        break
    idx += 1
else:
    l.append(val)

print(l)

[1, 2, 3, 10]


In [ ]:
l = [1, 2, 3]
val = 3

idx = 0
while idx < len(l):
    if l[idx] == val:
        break
    idx += 1
else:
    l.append(val)

print(l)

[1, 2, 3]




---



### 05 - Loop Break and Continue inside a Try...Except...Finally

Recall that in a `try` statement, the `finally` clause always runs:

In [ ]:
from six import b
a = 10
b = 1
try:
    a / b
except ZeroDivisionError:
    print('division by 0')
finally:
    print('this always executes')

this always executes


In [ ]:
from six import b
a = 10
b = 0
try:
    a / b
except ZeroDivisionError:
    print('division by 0')
finally:
    print('this always executes')

division by 0
this always executes


So, what happens when using a `try` statement within a `while` loop, and a `continue` or `break` statement is encountered?

In [ ]:
a = 0
b = 2

while a < 3:
    print('----------')
    a += 1
    b -=1
    try:
        res = a / b
    except ZeroDivisionError:
        print('{0}, {1} - division by 0'.format(a, b))
        res = 0
        continue
    finally:
        print('{0}, {1} - always executes'.format(a, b))
    print('{0}, {1} - main loop'.format(a, b))

----------
1, 1 - always executes
1, 1 - main loop
----------
2, 0 - division by 0
2, 0 - always executes
----------
3, -1 - always executes
3, -1 - main loop


As you can see in the above result, the finally code still executed, even though the current iteration was cut short with the continue statement.
<p>This works the same with a break statement:

In [ ]:
a = 0
b = 2

while a < 3:
    print('----------')
    a += 1
    b -=1
    try:
        res = a / b
    except ZeroDivisionError:
        print('{0}, {1} - division by 0'.format(a, b))
        res = 0
        break
    finally:
        print('{0}, {1} - always executes'.format(a, b))
    print('{0}, {1} - main loop'.format(a, b))

----------
1, 1 - always executes
1, 1 - main loop
----------
2, 0 - division by 0
2, 0 - always executes


We can even combine all this with the `else` clause:

In [ ]:
a = 0
b = 2

while a < 3:
    print('----------')
    a += 1
    b -=1
    try:
        res = a / b
    except ZeroDivisionError:
        print('{0}, {1} - division by 0'.format(a, b))
        res = 0
        break
    finally:
        print('{0}, {1} - always executes'.format(a, b))
    print('{0}, {1} - main loop'.format(a, b))
else:
    print('\n\nno errors were encountered!')

----------
1, 1 - always executes
1, 1 - main loop
----------
2, 0 - division by 0
2, 0 - always executes


In [ ]:
a = 0
b = 5

while a < 3:
    print('----------')
    a += 1
    b -=1
    try:
        res = a / b
    except ZeroDivisionError:
        print('{0}, {1} - division by 0'.format(a, b))
        res = 0
        break
    finally:
        print('{0}, {1} - always executes'.format(a, b))
    print('{0}, {1} - main loop'.format(a, b))
else:
    print('\n\nno errors were encountered!')

----------
1, 4 - always executes
1, 4 - main loop
----------
2, 3 - always executes
2, 3 - main loop
----------
3, 2 - always executes
3, 2 - main loop


no errors were encountered!


### 06 - The For Loop

In Python, an iterable is an object capable of returning values one at a time.

Many objects in Python are iterable: lists, strings, file objects and many more.



---

**BE CAREFUL!**

*Note: Our definition of an iterable did not state it was a collection of values - we only said it is an object that can return values one at a time - that's a subtle difference that when we look into iterators and generators.*



---



The `for` keyword can be used to iterate an iterable.

If you come with a background in another programming language, you have probably seen for loops defined this way:


```
for (int i=0; i < 5; i++) {
    //code block
}
```
This form of the `for` loop is simply a repetition, very similar to a `while` loop - in fact it is equivalent to what we could write in Python as follows:


In [ ]:
i = 0
while i < 5:
    #code block
    print(i)
    i += 1
i = None

0
1
2
3
4


But that's NOT what the `for` statement does in Python - the `for` statement is a way to iterate over iterables, and has nothing to do with the `for` loop we just saw. The closest equivalent we have in Python is the `while` loop written as above.

The use the `for` loop in Python, we require an iterable object to work with.

A simple iterable object is generated via the `range()` function

In [ ]:
for i in range(5):
    print(i)

0
1
2
3
4


Many objects are iterable in Python:

In [ ]:
for x in [1, 2, 3]:
    print(x)

1
2
3


In [ ]:
for x in 'hello':
    print(x)

h
e
l
l
o


In [ ]:
for x in ('a', 'b', 'c'):
    print(x)

a
b
c


When we iterate over an iterable, each iteration returns the `next` value (or object) in the iterable:

In [ ]:
for x in [(1, 2), (3, 4), (5, 6)]:
    print(x)

(1, 2)
(3, 4)
(5, 6)


We can even assign the individual tuple values to specific named variables:

In [ ]:
for i, j in [(1, 2), (3, 4), (5, 6)]:
    print(i, j)

1 2
3 4
5 6


The `break` and `continue` statements work just as well in `for` loops as they do in `while` loops:

In [ ]:
for i in range(5):
    if i == 3:
        continue
    print(i)

0
1
2
4


In [ ]:
for i in range(5):
    if i == 3:
        break
    print(i)

0
1
2




---
**BE CAREFUL**

*The `for` loop, like the `while` loop, also supports an `else` clause which is executed if and only if the loop terminates normally (i.e. did not exit because of a `break` statement)*

In [ ]:
for i in range(1, 5):
    print(i)
    if i % 7 == 0:
        print('multiple of 7 found')
        break
else:
    print('No multiples of 7 encountered')

1
2
3
4
No multiples of 7 encountered


In [ ]:
for i in range(1, 8):
    print(i)
    if i % 7 == 0:
        print('multiple of 7 found')
        break
else:
    print('No multiples of 7 encountered')

1
2
3
4
5
6
7
multiple of 7 found




---



Similarly to the `while` loop, `break` and `continue` work just the same in the context of a `try` statement's `finally` clause.

In [ ]:
for i in range(5):
    print('-----------')
    try:
        10 / (i - 3)
    except ZeroDivisionError:
        print('divided by 0')
        continue
    finally:
        print('always runs')
    print(i)

-----------
always runs
0
-----------
always runs
1
-----------
always runs
2
-----------
divided by 0
always runs
-----------
always runs
4


There are a number of standard techniques to iterate over iterables:

In [ ]:
s = 'hello'
for c in s:
    print(c)

h
e
l
l
o


But sometimes, for indexable iterable types (e.g. sequences), we want to also know the index of the item in the loop:

In [ ]:
s = 'hello'
i = 0
for c in s:
    print(i, c)
    i += 1

0 h
1 e
2 l
3 l
4 o


Slightly better approach might be:

In [ ]:
s = 'hello'

for i in range(len(s)):
    print(i, s[i])

0 h
1 e
2 l
3 l
4 o


or even better:

In [ ]:
s = 'hello'

for i, c in enumerate(s):
    print(i, c)

0 h
1 e
2 l
3 l
4 o


### 07 - Classes

**Custom Classes**

To create a custom class we use the `class` keyword, and we can initialize class attributes in the special method `__init__`.

In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height

We create instances of the `Rectangle` class by calling it with arguments that are passed to the `__init__` method as the second and third arguments. The first argument `(self)` is automatically filled in by Python and contains the object being created.



---
**BE CAREFUL!**


*Note that using `self` is just a convention (although a good one, and you should use it to make your code more understandable by others), you could really call it whatever (valid) name you choose.*



---



But just because you can does not mean you should!

In [ ]:
r1 = Rectangle(10, 20)
r2 = Rectangle(3, 5)

In [ ]:
r1.width

10

In [ ]:
r2.height

5

`width` and `height` are attributes of the `Rectangle` class. But since they are just values (not callables), we call them properties.

Attributes that are callables are called methods.

You'll note that we were able to retrieve the `width` and `height` attributes (properties) using a dot notation, where we specify the object we are interested in, then a dot, then the attribute we are interested in.

We can add callable attributes to our class (methods), that will also be referenced using the dot notation.

Again, we will create instance methods, which means the method will require the first argument to be the object being used when the method is called.

In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height
    
    def area(self):
        return self.width * self.height

    def perimeter(the_referenced_object):
        return 2 * (the_referenced_object.width + the_referenced_object.height)

In [ ]:
r1 = Rectangle(10, 20)

In [ ]:
r1.area()

200

When we ran the above line of code, our object was `r1`, so when `area` was called, Python in fact called the method area in the `Rectangle` class automatically passing `r1` to the `self` parameter.

This is why we can use a name other than `self`, such as in the `perimeter` method:

In [ ]:
r1.perimeter()

60

Again, I'm just illustrating a point, don't actually do that!

In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height
    
    def area(self):
        return self.width * self.height

    def perimeter(self):
        return 2 * (self.width + self.height)

In [ ]:
r1 = Rectangle(10, 20)

Python defines a bunch of special methods that we can use to give our classes functionality that resembles functionality of built-in and standard library objects.

Many people refer to them as magic methods, but there's nothing magical about them - unlike magic, they are well documented and understood!!

These special methods provide us an easy way to overload operators in Python.

For example, we can obtain the string representation of an integer using the built-in `str` function:

In [ ]:
str(10)

'10'

What happens if we try this with our `Rectangle` object?

In [ ]:
str(r1)

'<__main__.Rectangle object at 0x7fe790722350>'

Not exactly what we might have expected. On the other hand, how is Python supposed to know how to display our rectangle as a string?
<p>We could write a method in the class such as:

In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height
    
    def area(self):
        return self.width * self.height
    
    def perimeter(self):
        return 2 * (self.width + self.height)
    
    def to_str(self):
        return 'Rectangle (width={0}, height={1})'.format(self.width, self.height)

So we could get a string from our object as follows:

In [ ]:
r1 = Rectangle(10, 20)
r1.to_str()

'Rectangle (width=10, height=20)'

But of course, using the built-in `str` function still does not work:

In [ ]:
str(r1)

'<__main__.Rectangle object at 0x7fe790727f90>'

Does this mean we are our of luck, and anyone who writes a class in Python will need to provide some method to do this, and probably come up with their own name for the method too, maybe `to_str`, `make_string`, `stringify`, and who knows what else.

Fortunately, this is where these special methods come in. When we call `str(r1)`, Python will first look to see if our class `(Rectangle)` has a special method called  `__str__` .

If the `__str__` method is present, then Python will call it and return that value.

There's actually another one called `__repr__` which is related, but we'll focus on `__str__` for now.








In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        
    def area(self):
        return self.width * self.height
    
    def perimeter(self):
        return 2 * (self.width + self.height)
    
    def __str__(self):
        return 'Rectangle (width={0}, height={1})'.format(self.width, self.height)

In [ ]:
r1 = Rectangle(10, 20)

In [ ]:
str(r1)

'Rectangle (width=10, height=20)'

However, in Jupyter (and interactive console if you are using that), look what happens here:

In [ ]:
r1

As you can see we still get that default. That's because here Python is not converting `r1` to a string, but instead looking for a string representation of the object. It is looking for the `__repr__` method.

In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        
    def area(self):
        return self.width * self.height
    
    def perimeter(self):
        return 2 * (self.width + self.height)
    
    def __str__(self):
        return 'Rectangle (width={0}, height={1})'.format(self.width, self.height)
    
    def __repr__(self):
        return 'Rectangle({0}, {1})'.format(self.width, self.height)

In [ ]:
r1 = Rectangle(10, 20)

In [ ]:
print(r1) # uses __str__

Rectangle (width=10, height=20)


In [ ]:
r1 # uses __repr__

Rectangle(10, 20)

How about the comparison operators, such as `==` or `<` ?

In [ ]:
r1 = Rectangle(10, 20)
r2 = Rectangle(10, 20)

In [ ]:
r1 == r2

False

As you can see, Python does not consider `r1` and `r2` as equal (using the `==` operator). Again, how is Python supposed to know that two `Rectangle` objects with the same `height` and `width` should be considered equal?

We just need to tell Python how to do it, using the special method `__eq__`


In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        
    def area(self):
        return self.width * self.height
    
    def perimeter(self):
        return 2 * (self.width + self.height)
    
    def __str__(self):
        return 'Rectangle (width={0}, height={1})'.format(self.width, self.height)
    
    def __repr__(self):
        return 'Rectangle({0}, {1})'.format(self.width, self.height)
    
    def __eq__(self, other):
        print('self={0}, other={1}'.format(self, other))
        if isinstance(other, Rectangle):
            return (self.width, self.height) == (other.width, other.height)
        else:
            return False

In [ ]:
r1 = Rectangle(10, 20)
r2 = Rectangle(10, 20)

In [ ]:
r1 is r2

False

In [ ]:
r1 == r2

self=Rectangle (width=10, height=20), other=Rectangle (width=10, height=20)


True

In [ ]:
r3 = Rectangle(2, 3)

And if we try to compare our `Rectangle` to a different type:

In [ ]:
r1 == 100

self=Rectangle (width=10, height=20), other=100


False

Let's remove that `print` statement - I only put that in so you could see what the arguments were, in practice you should avoid side effects.

In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        
    def area(self):
        return self.width * self.height
    
    def perimeter(self):
        return 2 * (self.width + self.height)
    
    def __str__(self):
        return 'Rectangle (width={0}, height={1})'.format(self.width, self.height)
    
    def __repr__(self):
        return 'Rectangle({0}, {1})'.format(self.width, self.height)
    
    def __eq__(self, other):
        if isinstance(other, Rectangle):
            return (self.width, self.height) == (other.width, other.height)
        else:
            return False

What about `<`, `>`, `<=`, etc.?

Again, Python has special methods we can use to provide that functionality.

These are methods such as `__lt__`, `__gt__`, `__le__`, etc.

In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        
    def area(self):
        return self.width * self.height
    
    def perimeter(self):
        return 2 * (self.width + self.height)
    
    def __str__(self):
        return 'Rectangle (width={0}, height={1})'.format(self.width, self.height)
    
    def __repr__(self):
        return 'Rectangle({0}, {1})'.format(self.width, self.height)
    
    def __eq__(self, other):
        if isinstance(other, Rectangle):
            return (self.width, self.height) == (other.width, other.height)
        else:
            return False
    
    def __lt__(self, other):
        if isinstance(other, Rectangle):
            return self.area() < other.area()
        else:
            return NotImplemented

In [ ]:
r1 = Rectangle(100, 200)
r2 = Rectangle(10, 20)

In [ ]:
r1 < r2

False

In [ ]:
r2 < r1

True

What about `>` ?

In [ ]:
r1 > r2

True

How did that work? We did not define a `__gt__` method

Well, Python cleverly decided that since `r1 > r2` was not implemented, it would give `r2 < r1`

a try. And since, `__lt__` is defined, it worked!

Of course, `<=` is not going to magically work!

In [ ]:
r1 <= r2

TypeError: ignored

If you come from a Java background, you are probably thinking that using "bare" properties (direct access), such as `height` and `width` is a terrible design idea.

It is for Java, but not for Python.

Although you can use bare properties in Java, If you ever need to intercept the getting or setting of a property, you will need to write a method (such as `getWidth` and `setWidth`.) The problem is that if you used a  bare width property for example, a lot of your code might be using `obj.width` (as we have been doing here). The instant you make the `width` private and instead implement getters and setters, you break your code. Hence one of the reasons why in Java we just write getters and setters for properties from the beginning.

With Python this is not the case - we can change any bare property into getters and setters without beraking the code that uses that bare property.

I'll show you a quick example here, but we'll come back to this topic in much more detail later.
Let's take our `Rectangle` class once again. I'll use a simplified version to keep the code short.

In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height
    
    def __repr__(self):
        return 'Rectangle({0}, {1})'.format(self.width, self.height)

In [ ]:
r1 = Rectangle(10, 20)

In [ ]:
r1.width

10

In [ ]:
r1.width = 100

In [ ]:
r1

Rectangle(100, 20)

As you saw we can get and set the width property directly.

But let's say after this code has been released for a while and users of our class have been using it (and specifically setting and getting the `width` and `height` attribute a lot), but now we want to make sure users cannot set a non-positive value (i.e. `<=` 0) for `width` (or `height`, but we'll focus on `width` as an example).

In a language like Java, we would implement `getWidth` and `setWidth` and make width private - which would break any code directly accessing the `width` property.

In Python we can use some special decorators (more on those later) to encapsulate our property getters and setters:

In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self._width = width
        self._height = height
    
    def __repr__(self):
        return 'Rectangle({0}, {1})'.format(self.width, self.height)
    
    @property
    def width(self):
        return self._width
    
    @width.setter
    def width(self, width):
        if width <= 0:
            raise ValueError('Width must be positive.')
        self._width = width
    
    @property
    def height(self):
        return self._height
    
    @height.setter
    def height(self, height):
        if height <= 0:
            raise ValueError('Height must be positive.')
        self._height = height

In [ ]:
r1 = Rectangle(10, 20)

In [ ]:
r1.width

10

In [ ]:
r1.width = 100

In [ ]:
r1

Rectangle(100, 20)

In [ ]:
r1.width = -10

ValueError: ignored

There are more things we should do to properly implement all this, in particular we should also be checking the positive and negative values during the `__init__` phase. We do so by using the accessor methods for `height` and `width`:

In [ ]:
class Rectangle:
    def __init__(self, width, height):
        self._width = None
        self._height = None
        # now we call our accessor methods to set the width and height
        self.width = width
        self.height = height
    
    def __repr__(self):
        return 'Rectangle({0}, {1})'.format(self.width, self.height)
    
    @property
    def width(self):
        return self._width
    
    @width.setter
    def width(self, width):
        if width <= 0:
            raise ValueError('Width must be positive.')
        self._width = width
    
    @property
    def height(self):
        return self._height
    
    @height.setter
    def height(self, height):
        if height <= 0:
            raise ValueError('Height must be positive.')
        self._height = height

In [ ]:
r1 = Rectangle(0, 10)

ValueError: ignored

There more we should be doing, like checking that the width and height being passed in are numeric types, and so on. Espcially during the `__init__` phase. - we would rather raise an exception when the object is being created rather than delay things and raise an exception when the user calls some method like area - that way the exception will be on the line that creates the object - makes debugging much easier!

There are mant more of these special methods, and we'll look in detail at them later in this course.

## Section 03 - Variables and Memory

### 01 - Variables and Memory References

We can find the memory address that a variable references, by using `id()` function.

The `id()` function returns the memory address of its argument as a base-10 integer.

We can use the function `hex()` to convert the base-10 to base-16

In [ ]:
my_var = 10
print('my_var = {0}'.format(my_var))
print('memory address of my_var (decimal): {0}'.format(id(my_var)))
print('memory address of my_var (hex): {0}'.format(hex(id(my_var))))

my_var = 10
memory address of my_var (decimal): 11122880
memory address of my_var (hex): 0xa9b8c0


In [ ]:
greeting = 'Hello'
print('greeting = {0}'.format(greeting))
print('memory address of my_var (decimal): {0}'.format(id(greeting)))
print('memory address of my_var (hex): {0}'.format(hex(id(greeting))))

greeting = Hello
memory address of my_var (decimal): 139693754789808
memory address of my_var (hex): 0x7f0cfca1afb0




---
**BE CAREFUL!**


*Note how the memory address of `my_var` is different from that of `greeting`.*

*Strictly speaking, `my_var` is not "equal" to 10.*

*Instead `my_var` is a reference to an (integer) object (containing the value 10) located at the memory address `id(my_var)`*

*Similarly for the variable `greeting`*



---



### 02 - Reference Counting

Method that returns the reference count for a given variable's memory address:

In [ ]:
import ctypes

def ref_count(address):
    return ctypes.c_long.from_address(address).value

Let's make a variable, and check it's reference count:

In [ ]:
my_var = [1, 2, 3, 4]
ref_count(id(my_var))

1

There is another built-in function we can use to obtain the reference count:

In [ ]:
import sys
sys.getrefcount(my_var)

3

But why is this returning 3, instead of the expected 2 we obtained with the previous function?

Answer: The `sys.getrefcount()` function takes `my_var` as an argument, this means it receives (and stores) a reference to `my_var`'s memory address also - hence the count is off by 1. So we will use `from_address()` instead.

We make another reference to the same reference as `my_var`:

In [ ]:
other_var = my_var

Let's look at the memory address of those two variables and the reference counts:

In [ ]:
print(hex(id(my_var)), hex(id(other_var)))
print(ref_count(id(my_var)))

0x7f0cf72fa5f0 0x7f0cf72fa5f0
4


Force one reference to go away:

In [ ]:
other_var = None

And we look at the reference count again:

In [ ]:
print(ref_count(id(my_var)))

2


We see that the reference count has gone back to 2.

You'll probably never need  to do anythinli like this in Python. Memory management is completely transparent - this is just to illustrate some of what is going behind the scenes as it helps to understand upcoming concepts.



---
**BE CAREFUL!**


*In Google Colab Environment, `getrefcount()` function gives us the value 3, but in your local Jupyter notebook, it will be 2.*



---



### 03 - Garbage Collection

In [ ]:
import ctypes
import gc

We use the same function that we used in the lesson on reference counting to calculate the numver of references to a specified object (using its memory address to avoid creating an extra reference)

In [ ]:
def ref_count(address):
    return ctypes.c_long.from_address(address).value

We create a function that will search the objects in the GC for a specified id and tell us if the object was found or not:

In [ ]:
def object_by_id(object_id):
    for obj in gc.get_objects():
        if id(obj) == object_id:
            return "Object exists"
    return "Not found"

Next we define two classes that we will use to create a circular reference.

Class A's constructor will create an instance of class B and pass itself to class B's constructor that will then store that reference in some instance variable.

In [ ]:
class A:
    def __init__(self):
        self.b = B(self)
        print('A: self: {0}, b:{1}'.format(hex(id(self)), hex(id(self.b))))

In [ ]:
class B:
    def __init__(self, a):
        self.a = a
        print('B: self: {0}, a: {1}'.format(hex(id(self)), hex(id(self.a))))

We turn off the GC so we can see how reference counts are affected when the GC does not run and whne it does (by running it manually)

In [ ]:
gc.disable()

Now we create an instance of A, which will, in turn, create an instance of B which will store a reference to the calling A instance.

In [ ]:
my_var = A()

B: self: 0x7f0cf738f910, a: 0x7f0cf738f850
A: self: 0x7f0cf738f850, b:0x7f0cf738f910


As we can see A and B's constructors ran, and we also see from the memory addresses that we have a circular reference.

In fact `my_var` is also a reference to the same A instance:

In [ ]:
print(hex(id(my_var)))

0x7f0cf738f850


Another way to see this:

In [ ]:
print('a: \t{0}'.format(hex(id(my_var))))
print('a.b: \t{0}'.format(hex(id(my_var.b))))
print('b.a: \t{0}'.format(hex(id(my_var.b.a))))

a: 	0x7f0cf738f850
a.b: 	0x7f0cf738f910
b.a: 	0x7f0cf738f850


In [ ]:
a_id = id(my_var)
b_id = id(my_var.b)

We can see how many references we have for `a` and `b`

In [ ]:
my_var = None

In [ ]:
print('refcount(a) = {0}'.format(ref_count(a_id)))
print('refcount(b) = {0}'.format(ref_count(b_id)))
print('a: {0}'.format(object_by_id(a_id)))
print('b: {0}'.format(object_by_id(b_id)))

refcount(a) = 1
refcount(b) = 1
a: Object exists
b: Object exists


As we can see, the reference counts are now both equal to 1 (a pure circular reference), and reference counting alone did not destroy the A and B instances - they're still around. If no garbage collection is performed this would result in a memory leak.

Let's run the GC manually and re-check whether the objects still exist:

In [ ]:
gc.collect()
print('refcount(a) = {0}'.format(ref_count(a_id)))
print('refcount(b) = {0}'.format(ref_count(b_id)))
print('a: {0}'.format(object_by_id(a_id)))
print('b: {0}'.format(object_by_id(b_id)))

refcount(a) = 0
refcount(b) = 0
a: Not found
b: Not found


### 04 - Dynamic vs Static Typing

Python is dynamically typed.

This means that the type of a variable is simply the type of the object the variable name points to (references).The variable itself has no associated type.

In [ ]:
a = "hello"

In [ ]:
type(a)

str

In [ ]:
a = 10

In [ ]:
type(a)

int

In [ ]:
a = lambda x: x ** 2

In [ ]:
a(2)

4

In [ ]:
type(a)

function

As you can see from the above examples, the type of the variable `a` changed over time - in fact it was simply the type of the object `a` was referencing at that time. No type was ever attached to the variable name itself.

### 05 - Variable Re-Assignment

Notice how the memory address of `a` is different every time.

In [ ]:
a = 10
hex(id(a))

'0xa9b8c0'

In [ ]:
a = 15
hex(id(a))

'0xa9b960'

In [ ]:
a = 5
hex(id(a))

'0xa9b820'

In [ ]:
a = a + 1
hex(id(a))

'0xa9b840'

However, look at this:

In [ ]:
a = 10
b = 10
print(hex(id(a)))
print(hex(id(b)))

0xa9b8c0
0xa9b8c0


The memory adresses of both `a` and `b` are the same!!

### 06 - Object Mutability

Certain Python built-in object types (aka data types) are mutable.

That is, the internal contents (state) of the object in memory can be modified.

In [ ]:
my_list = [1, 2, 3]
print(my_list)
print(hex(id(my_list)))

[1, 2, 3]
0x7f0cf73abbe0


In [ ]:
my_list.append(4)
print(my_list)
print(hex(id(my_list)))

[1, 2, 3, 4]
0x7f0cf73abbe0


As you can see, the memory address of `my_list` has not changed.

But the contents of `my_list` has changed from [1, 2, 3] to [1, 2, 3, 4].




---
**BE CAREFUL!**


*On the other hand, consider this:*

In [ ]:
my_list_1 = [1, 2, 3]
print(my_list_1)
print(hex(id(my_list_1)))

[1, 2, 3]
0x7f0cef981f50


In [ ]:
my_list_1 = my_list_1 + [4]
print(my_list_1)
print(hex(id(my_list_1)))

[1, 2, 3, 4]
0x7f0cef988aa0


*Notice here that the memory address of `my_list_1` did change.*

*This is because concatenating two lists objects `my_list_1` and `[4]` did not modify the contents of `my_list_1` - instead it created a new list object and re-assigned `my_list_1` to reference this new object.*

*Similarly with ditionary objects that are also mutable types.*



---



In [ ]:
my_dict = dict(key1 = 'value 1')
print(my_dict)
print(hex(id(my_dict)))

{'key1': 'value 1'}
0x7f0cef8f80a0


In [ ]:
my_dict['key1'] = 'modified value 1'
print(my_dict)
print(hex(id(my_dict)))

{'key1': 'modified value 1'}
0x7f0cef8f80a0


In [ ]:
my_dict['key2'] = 'value 2'
print(my_dict)
print(hex(id(my_dict)))

{'key1': 'modified value 1', 'key2': 'value 2'}
0x7f0cef8f80a0


Once again we see that while we are modifying the contents of the dictionary, the memory address of `my_dict` has not changed.

Now consider the immutable sequence type: tuple

The tuple is immutable, so elements cannot be added, removed or replaced.

In [ ]:
t = (1 ,2 ,3)

This tuple will never change at all. It has three elementes, the integers 1, 2 , and 3. This will remain the case as long as `t`'s reference is not changed.



---
**BE CAREFUL!**


*But, consider the following tuple:*

In [ ]:
a = [1, 2]
b = [3, 4]
t = (a, b)

*Now, `t` is still immutable, i.e. it contains a reference to the object `a` abd the object `b`. That will never change as long as `t`'s reference is not re-assigned.*

*However, the elements `a` and `b` are, themselves, mutable.*

In [ ]:
a.append(3)
b.append(5)
print(t)

([1, 2, 3], [3, 4, 5])


*Observe that the contents of `a` and `b` did change!*

*So immutability can be a littlme more subtle than just thinking something can never change.*

*The tuple `t` did not change - it contains two elements, that are the references `a` and `b`. And that will not change. But, because the referenced elements are mutable themselves, it appears as though the tuple has changed.*

*It  hasn't though - tha distinction is subtle but important to understand!*



---



### 07 - Function Arguments and Mutability

Consider a function that receives a string argument, and changes the argument in some way:

In [ ]:
def process(s):
    print('initial s # = {0}'.format(hex(id(s))))
    s = s + ' world'
    print('s after change # = {0}'.format(hex(id(s))))

In [ ]:
my_var = 'hello'
print('my_var # = {0}'.format(hex(id(my_var))))

my_var # = 0x7f0cfc68fa30


Note that when s is received, it is referencing the same object as `my_var`.

After we "modify" s, s is pointing to a new memory address:

In [ ]:
process(my_var)

initial s # = 0x7f0cfc68fa30
s after change # = 0x7f0cef530db0


And our own variable `my_var` is still pointing to the original memory address:

In [ ]:
print('my_var # = {0}'.format(hex(id(my_var))))

my_var # = 0x7f0cfc68fa30


Let's see how this works with mutable objects:

In [ ]:
def modify_list(items):
    print('initial items # = {0}'.format(hex(id(items))))
    if len(items) > 0:
        items[0] = items[0] ** 2
    items.pop()
    items.append(5)
    print('final items # = {0}'.format(hex(id(items))))

In [ ]:
my_list = [2, 3, 4]
print('my_list # = {0}'.format(hex(id(my_list))))

my_list # = 0x7f0cef388f50


In [ ]:
modify_list(my_list)

initial items # = 0x7f0cef388f50
final items # = 0x7f0cef388f50


In [ ]:
print(my_list)
print('my_list # = {0}'.format(hex(id(my_list))))

[4, 3, 5]
my_list # = 0x7f0cef388f50


As you can see, thoughout all the code, the memory address referenced by `my_list` and `items` is always the same (shared) reference - we are simply modifying the contents (internal state) of the object at that memory address.

Now, even with immutable container objects we have to be careful, e.g. a tuple containing a list (the tuple is immutable, but the list element inside the tuple is mutable)

In [ ]:
def modify_tuple(t):
    print('initial t # = {0}'.format(hex(id(t))))
    t[0].append(100)
    print('final t # = {0}'.format(hex(id(t))))

In [ ]:
my_tuple = ([1, 2], 'a')

In [ ]:
hex(id(my_tuple))

'0x7f0cef29b410'

In [ ]:
modify_tuple(my_tuple)

initial t # = 0x7f0cef29b410
final t # = 0x7f0cef29b410


In [ ]:
my_tuple

([1, 2, 100], 'a')

As you can see, the first element of the tuple was mutated.

### 08 - Shared References and Mutability

The following sets up a shared reference between the variables `my_var_1` and `my_var_2`

In [1]:
my_var_1 = 'hello'
my_var_2 = my_var_1
print(my_var_1)
print(my_var_2)

hello
hello


In [2]:
print(hex(id(my_var_1)))
print(hex(id(my_var_2)))

0x7f52d07a7d70
0x7f52d07a7d70


In [3]:
my_var_2 = my_var_2 + ' world!'

In [4]:
print(hex(id(my_var_1)))
print(hex(id(my_var_2)))

0x7f52d07a7d70
0x7f52cb4cb8b0




---
**BE CAREFUL!**


*Be careful if the variable type is mutable!*

*Here we create a list `(my_list_1)` and create a variable `(my_list_2)` referencing the same list object:*

In [5]:
my_list_1 = [1, 2, 3]
my_list_2 = my_list_1
print(my_list_1)
print(my_list_2)

[1, 2, 3]
[1, 2, 3]


*As we can see they have the same memory address (shared reference):*

In [6]:
print(hex(id(my_list_1)))
print(hex(id(my_list_2)))

0x7f52cb46ea00
0x7f52cb46ea00




---



Now we modify the list referenced by `my_list_2`:

In [7]:
my_list_2.append(4)

`my_list_2` has been modified:

In [8]:
print(my_list_2)

[1, 2, 3, 4]


And since my_list_1 references the same list object, it has also changed:

In [9]:
print(my_list_1)

[1, 2, 3, 4]


As you can see, both variables still share the same reference:

In [10]:
print(hex(id(my_list_1)))
print(hex(id(my_list_2)))

0x7f52cb46ea00
0x7f52cb46ea00


**Behind the scenes with Python's memory manager**

Recall from a few lectures back:

In [11]:
a = 10
b = 10

In [12]:
print(hex(id(a)))
print(hex(id(b)))

0xa9b8c0
0xa9b8c0


Same memory address!!

This is safe for Python to do because integer objects are immutable.

So, even though `a` and `b` initially shared the same memory address, we can never modify a's value by "modifying" b' value.

The only way to change `b`'s value is to change it's reference, which will never affect `a`.

In [13]:
b = 15

In [14]:
print(hex(id(a)))
print(hex(id(b)))

0xa9b8c0
0xa9b960


However, for mutable objects, Python's memory manager does not do this, since that would not be safe.

In [15]:
my_list_1 = [1, 2, 3]
my_list_2 = [1, 2 ,3]

As you can see, although the two variables were assingend identical "contents", the memory addresses are not the same:

In [17]:
print(hex(id(my_list_1)))
print(hex(id(my_list_2)))

0x7f52c3ccd780
0x7f52c3d3ceb0


### 09 - Variable Equality

From the previous lecture we know that `a` and `b` will have a shared reference:

In [18]:
a = 10
b = 10

print(hex(id(a)))
print(hex(id(b)))

0xa9b8c0
0xa9b8c0


When we use the `is` operator, we are comparing t he moemory address references:

In [19]:
print("a is b ", a is b)

a is b  True


But if we use the `==` operator, we are comparing the contents:

In [20]:
print("a == b", a == b)

a == b True


The following however, do not have a shared reference:

In [21]:
a = [1, 2, 3]
b = [1, 2, 3]

print(hex(id(a)))
print(hex(id(b)))

0x7f52c3d2d500
0x7f52c3d2d280


Although they are not the same objects, they do contain the same "values"

In [22]:
print("a is b", a is b)
print("a == b", a == b)

a is b False
a == b True


Python will attempt to compare values as best as posssible, for example:

In [23]:
a = 10
b = 10.0

These are not the same reference, since one object is an `int` and the other is a `float`

In [24]:
print(type(a))
print(type(b))

<class 'int'>
<class 'float'>


In [25]:
print(hex(id(a)))
print(hex(id(b)))

0xa9b8c0
0x7f52c3d19510


In [26]:
print("a is b", a is b)
print("a == b", a == b)

a is b False
a == b True


So, even though `a` is an integer 10, and `b` is a float 10.0, the values will still compare as equal.

In fact, this will also have the same behavior:

In [27]:
c = 10 + 0j
print(type(c))

<class 'complex'>


In [28]:
print("a is c", a is c)
print("a == c", a == c)

a is c False
a == c True


**The None Object**

`None` is a built-in "variable" of type `NoneType`.

Basically the keyword `None` is a reference to an object instance of `NoneType`.

NoneType objects are immutable! Python's memory manager will therefore use shared references to the None object.

In [29]:
print(None)

None


In [30]:
hex(id(None))

'0xa9ae00'

In [32]:
type(None)

NoneType

In [31]:
a = None
print(type(a))
print(hex(id(a)))

<class 'NoneType'>
0xa9ae00


In [33]:
a is None

True

In [34]:
a == None

True

In [36]:
b = None
hex(id(b))

'0xa9ae00'

In [37]:
a is b

True

In [38]:
a == b

True

In [39]:
l = []

In [40]:
type(l)

list

In [41]:
l is None

False

In [42]:
l == None

False

### 10 - Everything is an Object

In [43]:
a = 10

`a` is an object of type `int`, i.e. `a` is an instance of the `int` class.

In [44]:
print(type(a))

<class 'int'>


if `int` is a class, we should be able to declare it using standard class instatiation:

In [45]:
b = int(10)

In [47]:
print(b)
print(type(b))

10
<class 'int'>


We can even request the class documentation:

In [48]:
help(int)

Help on class int in module builtins:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __ceil__(...)
 |      Ceiling of

As we see from the docs, we can even create an `int` using an overloaded constructor:

In [49]:
b = int('10', base=2)

In [50]:
print(b)
print(type(b))

2
<class 'int'>


**Functions are Objects too**

In [51]:
def square(a):
    return a ** 2

In [52]:
type(square)

function

In fact, we can even assign them to a variable:

In [53]:
f = square

In [54]:
type(f)

function

In [55]:
f is square

True

In [56]:
f(2)

4

In [57]:
type(f(2))

int

A function can return a function:

In [58]:
def cube(a):
    return a ** 3

In [59]:
def select_function(fn_id):
    if fn_id == 1:
        return square
    else:
        return cube

In [60]:
f = select_function(1)
print(hex(id(f)))
print(hex(id(square)))
print(hex(id(cube)))
print(type(f))
print('f is square: ', f is square)
print('f is cube: ', f is cube)
print(f)
print(f(2))

0x7f52c3bdb560
0x7f52c3bdb560
0x7f52c3bdb440
<class 'function'>
f is square:  True
f is cube:  False
<function square at 0x7f52c3bdb560>
4


We could even call it this way:

In [61]:
select_function(1)(5)

25

A Function can be passed as an argument to another function.

(This example is pretty useless, but it illustrates the point effectively)

In [62]:
def exec_function(fn, n):
    return fn(n)

In [63]:
result = exec_function(cube, 2)
print(result)

8


### 11 - Python Optimizations - Interning

Earlier, we saw shared references being created automatically by Python:

In [64]:
a = 100
b = 100
print(id(a))
print(id(b))

11125760
11125760


Note how `a` and `b` reference the same object.



---
**BE CAREFUL!**

*But consider the following example:*

In [65]:
a = 500
b = 500
print(id(a))
print(id(b))

139993448788304
139993448791536


*As you can see, the variables `a` and `b` do not point to the same object!*

*This is because Python pre-caches integer objects in the range [-5, 256]*

*So for example:*

In [66]:
a = 256
b = 256
print(id(a))
print(id(b))

11130752
11130752


*and*

In [67]:
a = -5
b = -5
print(id(a))
print(id(b))

11122400
11122400


do have the same reference.

*This is called **interning**: Python **interns** the integers in the range [-5, 256].*

*The integers in the range [-5, 256] are essentially **singleton** objects.*

In [68]:
a = 10
b = int(10)
c = int('10')
d = int('1010', 2)

In [69]:
print(a, b, c, d)

10 10 10 10


In [70]:
a is b

True

In [71]:
a is c

True

In [72]:
a is d

True

As you can see, all these variables were created in different ways, but since the integer object with value 10 behaves like a singleton, they all ended up pointing to the same object in memory.

### 12 - Python Optimizations: String Interning

Python will automatically intern certain strings.

In particular all the identifiers (variable names, function names, class names, etc) are interned (singleton objects created).

Python will also intern string literals that look like identifiers.

For example:

In [73]:
a = 'hello'
b = 'hello'
print(id(a))
print(id(b))

139993661734256
139993661734256




---
**BE CAREFUL!**


*But not the following:*

In [74]:
a = 'hello, world!'
b = 'hello, world!'
print(id(a))
print(id(b))

139993448058480
139993448057968




---



However, because the following literals resemble identifiers, even though they are quite long, Python will still automatically intern them:

In [75]:
a = 'hello_world'
b = 'hello_world'
print(id(a))
print(id(b))

139993448058160
139993448058160


And even longer:

In [76]:
a = '_this_is_a_long_string_that_could_be_used_as_an_identifier'
b = '_this_is_a_long_string_that_could_be_used_as_an_identifier'
print(id(a))
print(id(b))

139993448183824
139993448183824


Even if the string starts with a digit:

In [77]:
a = '1_hello_world'
b = '1_hello_world'
print(id(a))
print(id(b))

139993448052016
139993448052016


That was interned (pointer is the same), but look at this one:

In [78]:
a = '1 hello world'
b = '1 hello world'
print(id(a))
print(id(b))

139993448229680
139993448229808


Interning strings (making them singleton objects) means that testing for string equality can be done faster by comparing the memory address:

In [79]:
a = 'this_is_a_long_string'
b = 'this_is_a_long_string'
print('a==b:', a == b)
print('a is b:', a is b)

a==b: True
a is b: True




---
**BE CAREFUL!**


*Note: Remember, using `is` ONLY works if the strings were interned!*

*Here's where this technique fails:*

In [80]:
a = 'hello world'
b = 'hello world'
print('a==b:', a==b)
print('a is b:', a is b)

a==b: True
a is b: False




---



You can force strings to be interned (but only use it if you have a valid performance optimization need):

In [81]:
import sys

In [82]:
a = sys.intern('hello world')
b = sys.intern('hello world')
c = 'hello world'
print(id(a))
print(id(b))
print(id(c))

139993448083824
139993448083824
139993448083568


Notice how `a` and `b` are pointing to the same object, but `c` is NOT.

So, since both `a` and `b` were interned we can use `is` to test for equality of the two strings:

In [84]:
print('a==b:', a==b)
print('a is b', a is b)

a==b: True
a is b True


So, does interning really make a big speed difference?

Yes, but only if you are performing a lot of comparisons.

Let's run som quick and dirty benchamarks:

In [85]:
def compare_using_equals(n):
    a = 'a long string that is not interned' * 200
    b = 'a long string that is not interned' * 200
    for i in range(n):
        if a == b:
            pass

In [86]:
def compare_using_interning(n):
    a = sys.intern('a long string that is not interned' * 200)
    b = sys.intern('a long string that is not interned' * 200)
    for i in range(n):
        if a is b:
            pass

In [87]:

import time

start = time.perf_counter()
compare_using_equals(10000000)
end = time.perf_counter()

print('equality: ', end-start)

equality:  1.61897275299998


In [88]:
start = time.perf_counter()
compare_using_interning(10000000)
end = time.perf_counter()

print('identity: ', end-start)

identity:  0.49664921299972775


As you can see, the performance difference, especially for long strings, and for many comparisons, can be quite radical!

### 13 - Python Optimizations - PeepHole

Peephole optimizations refer to a certain class of optimization stratigies Python employs during any compilation phases.

**Constant Expressions**

Let's see how Python reduces constant expressions for optimization purposes:

In [89]:
def my_func():
    a = 24 *60
    b = (1, 2) * 5
    c = 'abc' * 3
    d = 'ab' * 11
    e = 'the quick brown fox' * 10
    f = [1, 2] * 5

In [91]:
my_func.__code__.co_consts

(None,
 1440,
 (1, 2, 1, 2, 1, 2, 1, 2, 1, 2),
 'abcabcabc',
 'ababababababababababab',
 'the quick brown foxthe quick brown foxthe quick brown foxthe quick brown foxthe quick brown foxthe quick brown foxthe quick brown foxthe quick brown foxthe quick brown foxthe quick brown fox',
 1,
 2,
 5)

As you can see in the example above `24 * 60` was pre-calculated and cached as a constant (`1440`).

Similarly, `(1, 2) * 5` was cached as `(1, 2, 1, 2, 1, 2, 1, 2, 1, 2)` and `'abc' * 3` was cached as `abcabcabc`

On the other hand, note how `'the quick brown fox'* 10` was not pre-calculated (too long).

Similarly `[1, 2] * 5` was not pre-calculated either siince a list is mutable, and hence not a constant.

**Membership Tests**

In membership testing, optimizations are applied as can be seen below:

In [92]:
def my_func():
    if e in [1, 2, 3]:
        pass

In [93]:
my_func.__code__.co_consts

(None, (1, 2, 3))

As you can see, the mutable list `[1, 2, 3]` was converted to an immutable tuple.

It is OK to do this here, since we are testing membersip of the list at that point in time, hence it is safe to convert it to a tuple, which is more efficient tan testing membership of a list.

In the same way, set membership will be converted to frozen set membership:

In [96]:
def my_func():
    if e in {1, 2, 3}:
        pass

In [97]:
my_func.__code__.co_consts

(None, frozenset({1, 2, 3}))

In general, when you are writing your code, if you can use `set` membership testing, prefer that over a list or tuple - it is quite a bit more efficien.

Let's do a small quick (and dirty) benchmark of this:

In [98]:
import string
import time

char_list = list(string.ascii_letters)
char_tuple = tuple(string.ascii_letters)
char_set = set(string.ascii_letters)

print(char_list)
print()
print(char_tuple)
print()
print(char_set)

['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']

('a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z')

{'U', 't', 'V', 'C', 's', 'Z', 'A', 'q', 'y', 'n', 'e', 'G', 'S', 'T', 'W', 'Y', 'b', 'Q', 'l', 'a', 'u', 'c', 'J', 'k', 'm', 'r', 'x', 'g', 'O', 'p', 'F', 'o', 'E', 'M', 'v', 'B', 'P', 'L', 'd', 'H', 'X', 'R', 'z', 'h', 'f', 'i', 'I', 'D', 'N', 'w', 'K', 'j'}


In [99]:
def membership_test(n, container):
    for i in range(n):
        if 'p' in container:
            pass

In [100]:
start = time.perf_counter()
membership_test(10000000, char_list)
end = time.perf_counter()
print('list membership: ', end-start)

list membership:  3.218021790000421


In [101]:
start = time.perf_counter()
membership_test(10000000, char_tuple)
end = time.perf_counter()
print('tuple membership: ', end-start)

tuple membership:  3.319882019000943


In [102]:
start = time.perf_counter()
membership_test(10000000, char_set)
end = time.perf_counter()
print('set membership: ', end-start)

set membership:  0.6080441869999049


As you can see, set membership tests run quite a bit faster - which is not suprising since they are basically dictionary-like objects, so hash maps are used for looking up an item to determine membership.